util/types/datum_test.go

```diff
@@ -366,3 +368,205 @@ func (ts *testDatumSuite) TestToBytes(c *C) {
 		c.Assert(bin, BytesEquals, tt.out)
 	}
 }
+
+func mustParseDurationDatum(str string, fsp int) Datum {
+	dur, err := ParseDuration(str, fsp)
+	if err != nil {
+		panic(err)
+	}
+	return NewDurationDatum(dur)
+}
+
+func mustParseHexDatum(str string) Datum {
+	hex, err := ParseHex(str)
+	if err != nil {
+		panic(err)
+	}
+	var d Datum
+	d.SetMysqlHex(hex)
+	return d
+}
+
+func (ts *testDatumSuite) TestCoerceArithmetic(c *C) {
+	sc := &variable.StatementContext{TimeZone: time.UTC}
+	tests := []struct {
+		input  Datum
+		expect Datum
+		hasErr bool
+	}{
+		{NewStringDatum("12.5"), NewFloat64Datum(12.5), false},
+		{NewStringDatum("asdf"), Datum{}, true},
+		{NewBytesDatum([]byte("12.527")), NewFloat64Datum(12.527), false},
+		{mustParseTimeIntoDatum("2017-07-18 17:21:42.321", mysql.TypeTimestamp, 3), NewDatum(NewDecFromStringForTest("20170718172142.321")), false},
+		{mustParseTimeIntoDatum("2017-07-18 17:21:42.32172", mysql.TypeDatetime, 0), NewIntDatum(20170718172142), false},
+		{mustParseDurationDatum("10:10:10", 0), NewIntDatum(101010), false},
+		{mustParseDurationDatum("10:10:10.100", 3), NewDatum(NewDecFromStringForTest("101010.100")), false},
+		{mustParseHexDatum("x'4D7953514C'"), NewFloat64Datum(332747985228), false},
+		{NewDatum(Bit{Value: 1, Width: 8}), NewFloat64Datum(1), false},
+		{NewDatum(Enum{"xxx", 1}), NewFloat64Datum(1), false},
+		{NewDatum(Set{"xxx", 1}), NewFloat64Datum(1), false},
+		{NewIntDatum(5), NewIntDatum(5), false},
+		{NewFloat64Datum(5.5), NewFloat64Datum(5.5), false},
+	}
+
+	for _, tt := range tests {
+		got, err := CoerceArithmetic(sc, tt.input)
+		c.Assert(err != nil, Equals, tt.hasErr)
+		v, err := got.CompareDatum(sc, tt.expect)
+		c.Assert(err, IsNil)
+		c.Assert(v, Equals, 0, Commentf("got:%#v, expect:%#v", got, tt.expect))
+	}
+}
+
+func (ts *testDatumSuite) TestComputePlusAndMinus(c *C) {
+	sc := &variable.StatementContext{TimeZone: time.UTC}
+	tests := []struct {
+		a      Datum
+		b      Datum
+		plus   Datum
+		minus  Datum
+		hasErr bool
+	}{
+		{NewIntDatum(72), NewIntDatum(28), NewIntDatum(100), NewIntDatum(44), false},
+		{NewIntDatum(72), NewUintDatum(28), NewIntDatum(100), NewIntDatum(44), false},
+		{NewUintDatum(72), NewUintDatum(28), NewUintDatum(100), NewUintDatum(44), false},
+		{NewUintDatum(72), NewIntDatum(28), NewUintDatum(100), NewUintDatum(44), false},
+		{NewFloat64Datum(72.0), NewFloat64Datum(28.0), NewFloat64Datum(100.0), NewFloat64Datum(44.0), false},
+		{NewDecimalDatum(NewDecFromStringForTest("72.5")), NewDecimalDatum(NewDecFromInt(3)), NewDecimalDatum(NewDecFromStringForTest("75.5")), NewDecimalDatum(NewDecFromStringForTest("69.5")), false},
+		{NewIntDatum(72), NewFloat64Datum(42), Datum{}, Datum{}, true},
```

Comment:
```Invalid operation: %v %v %v (mismatched types %T and %T)```
`ComputePlus` and `ComputeMinus` don't handle different types except int/uint, and don't cast input to decimal.